File: python/Sorting/countSort_v3.py
import sys
sys.stdin = open("input.txt","r")

def count_sort(arr):
    n = len(arr)
    if n<=1: return arr
    # find the minimum and the maximum from the array
    mn = min(arr)
    mx = max(arr)

    N = mx-mn+1 # find the length of the 'counting range from MIN to MAX'
    count = [0]*N
    for x in arr: count[x-mn]+=1 # count the frequency of the occurrence
    for i in range(1,N): count[i]+=count[i-1] # get the coumulative sum

    sorted_arr = [0]*n
    for x in arr:
        count[x-mn] -= 1
        ndx = count[x-mn] # find the position of the x where it should be placed
        sorted_arr[ndx] = x # place the number at the correct position
    
    return sorted_arr

arr = list(map(int,input().split()))
sorted_arr = count_sort(arr)
print(sorted_arr)

File: cpp/Sorting/countSort_v3.cpp
#include <iostream>
#include <vector>
using namespace std;

#define vi vector<int>

vi count_sort(vi &arr){
    int n = arr.size();
    if(n<=1) return arr;
    int MIN,MAX;
    MIN=MAX=arr[0];
    for(int x:arr){ // find the minimum and the maximum from the array
        MIN=min(MIN,x);
        MAX=max(MAX,x);
    }

    int N = MAX-MIN+1; // find the length of the 'counting range from MIN to MAX'
    vi count(N);
    for(int x:arr) ++count[x-MIN]; // count the frequency of the occurrence
    for(int i=1;i<N;i++) count[i]+=count[i-1]; // get the coumulative sum
    vi sorted_arr(n); 
    for(int x:arr){
        int ndx = --count[x-MIN]; // find the position of the x where it should be placed
        sorted_arr[ndx] = x; // place the number at the correct position
    }
    return sorted_arr;
}

int main(){
    freopen("input.txt","r",stdin);
    int n;
    cin >> n;
    vi arr(n);
    for(int i=0;i<n;i++) cin >> arr[i];
    vi sorted_arr = count_sort(arr);
    for(int x:sorted_arr) cout << x << ' ';
    cout << '\n';
}